lib/index.js

```diff
@@ -12,34 +26,308 @@ class PluginPassportOAuth {
+        });
+      });
+  }
+
+  initStrategies () {
+    this.strategies = {};
+    Object.keys(this.config.strategies).forEach(strategy => {
+      this.strategies[strategy] = {
+        config: {
+          constructor: this.strategyConstructor,
+          strategyOptions: this.config.strategies[strategy].credentials,
+          authenticateOptions: {
+            scope: this.config.strategies[strategy].scope
+          },
+          fields: ['access_token', 'refresh_token'],
+          verify: 'verify'
+        },
+        methods: {
+          exists: 'exists',
+          validate: 'validate',
+          create: 'create',
+          update: 'update',
+          delete: 'delete',
+          getInfo: 'getInfo'
+        }
+      };
+    });
+  }
+
+  /**
+   * Get the repository according to the strategy
+   *
+   * @param provider
+   * @returns {module.exports.constructors.Repository|*}
+   */
+  getProviderRepository (provider) {
+    if (!this.providerRepository) {
+      this.providerRepository = new this.context.constructors.Repository(provider);
     }
 
-    strategy = new Strategy(this.context, this.config);
-    strategy.init();
+    return this.providerRepository;
+  }
+
+  /**
+   * Get the credentials of a given userId from the repository of the corresponding strategy
+   *
+   * @param userId
+   * @param provider
+   * @returns {Promise|Promise.<TResult>}
+   */
+  getCredentialsFromUserId (userId, provider) {
+    return this.getProviderRepository(provider).search({
+      query: {
+        match: {
+          userId
+        }
+      }
+    })
+      .then(result => {
+        if (result.total === 0) {
+          return Promise.resolve(null);
+        }
+
+        return Promise.resolve(result.hits[0]);
+      });
+  }
+
+  /**
+   * Verify function called by passport to log the user.
+   * If a persist option is in the configuration of the strategy then every field present in this configuration will be persisted in the repository.
+   *
+   * @param request
+   * @param accessToken
+   * @param refreshToken
+   * @param profile
+   * @returns {Promise.<T>}
+   */
+  verify (request, accessToken, refreshToken, profile) {
+    const
+      user = {};
+
+    return this.getProviderRepository(profile.provider)
+      .get(profile._json[this.config.strategies[profile.provider].identifierAttribute])
+      .then(userObject => {
+        let req = null;
+
+        if (userObject !== null) {
+          return Promise.resolve(userObject.userId);
+        }
+
+        Object.keys(profile._json).forEach(attr => {
+          if (this.config.strategies[profile.provider].persist.indexOf(attr) > -1) {
+            user[attr] = profile._json[attr];
+          }
+        });
+
+        if (this.config.strategies[profile.provider].persist.length !== 0) {
+          req = {
+            controller: 'security',
+            action: 'createUser',
+            _id: profile._json[this.config.strategies[profile.provider].identifierAttribute],
```

Comment:
It's not necessary that Kuzzle user's `_id` is ewual to current strategy's identifierAttribute.
To avoid conflicts, `_id` should be autogenerated.